### PR TITLE
Grid support for lforms v29

### PR DESCRIPTION
**Pull request by nikhil-ht**
**Overview**
- Added grid layout support to lform

**How it was tested**

- Tested by rendering lform with grid layout 
- Tested by creating local  npm package with lforms build and using its in sapphire 

**Screenshot**
image

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,7 +2,6 @@ node_modules
 dist
 bower-dist/*.map
 bower_components
-/.*
```
Comment (hankwallace) [inline .gitignore:5]: Why was this removed?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,14 @@
+
```
Comment (hankwallace) [inline Dockerfile:1]: This is a library. Why is there a `Dockerfile` for it?

**Review by hankwallace — COMMENTED** [review]
A few comments, but please get Colin and/or Darshan to review.

Should we define a Dockerfile just for the purpose of building and pushing the artifact?

File: .gitignore (hunk 1 above)
Comment (nikhil-ht) [reply .gitignore:5]: @hankwallace this ignores all files that start with a (dot), so we need a .npmrc file for getting the npm token.

File: Dockerfile (hunk 2 above)
Comment (nikhil-ht) [reply Dockerfile:1]: @hankwallace, when publishing lform on npm we need to build lform, and lform requires a node 10 version, but in CICD jobs we cannot be able to change the node version. Therefore, I use a Docker file for building the form

File: .gitignore (hunk 1 above)
Comment (hankwallace) [reply .gitignore:5]: @nikhil-ht why not add a `!.npmrc` then like the lines below?

File: Dockerfile (hunk 2 above)
Comment (hankwallace) [reply Dockerfile:1]: That makes sense. It's unfortunate that we have to add the `Dockerfile` here though. 

@semicolin @ddjain can you think of a better way?

File: .gitignore (hunk 1 above)
Comment (nikhil-ht) [reply .gitignore:5]: @hankwallace  will add   !.npmrc

File: Dockerfile (hunk 2 above)
Comment (semicolin) [reply Dockerfile:1]: Maybe a GitHub workflow?

File: package.json [hunk 3]
```diff
@@ -89,7 +92,7 @@
     "build:clean": "rm -rf dist bower-dist app/generated .tmp",
     "build:mkdist": "mkdir dist; mkdir -p dist/$npm_package_name-$npm_package_version/styles; (cd dist; ln -s $npm_package_name-$npm_package_version latest)",
     "build:version": "node -e 'require(\"fs\").writeFileSync(\"app/version.json\", JSON.stringify({lformsVersion: require(\"./package.json\").version}))'",
-    "build:css:lforms": "cleancss --skip-rebase -o dist/latest/styles/lforms.min.css dist/latest/styles/lforms.css",
+    "build:css:lforms": "cleancss --skip-rebase -o dist/latest/styles/lforms.min.css dist/lforms-0.0.3/styles/lforms.css",
```
Comment (semicolin) [inline package.json]: why was this changed? we would need to update this every time we increment the version, right?

File: package.json [hunk 4]
```diff
@@ -98,5 +101,14 @@
...
+  },
+  "description": "[LForms](http://lhncbc.github.io/lforms/), a.k.a. LHC-Forms, is a feature-rich, open-source widget that creates input forms, based on definition files, for Web-based applications.  In addition to its native form-definition format, it partially supports the HL7 FHIR Questionnaire standard (SDC profile), and work is in progress to expand that support.",
+  "bugs": {
+    "url": "https://github.com/lhncbc/lforms/issues"
+  },
+  "homepage": "https://github.com/lhncbc/lforms#readme",
+  "directories": {
+    "test": "test"
+  },
+  "author": "Elimu"
```
Comment (semicolin) [inline package.json]: We might want to update the description etc. to make it clear that this is a fork of an old version of lforms that adds one specific feature and we are not the original authors of the library.

Comment (semicolin) [reply package.json]: I'm not familiar with forking etiquette. Should the URLs for git repo, bugs, and homepage point to our fork or to the original library?

File: app/views/partials/layout-grid.html [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+                aria-labelledby="{{lfData._horizontalTableInfo[item._horizontalTableId].columnHeaders[$index].id}}"
+                ng-focus="setActiveRow(subItem)" ng-blur="activeRowOnBlur(subItem)">
+
+            </div>
+          </ng-form>
+        </td>
+
+
+        <!-- last row data -->
+        <td ng-repeat="cell in subItem.items" ng-if="targetShown(cell) && !cell._isHiddenFromView"
```
Comment (semicolin) [inline app/views/partials/layout-grid.html]: The "last row" comment seems misleading. This is really creating an additional column for each nested item, right? Are there ever situations where each item has a different number of nested items, resulting in different number of columns in each row?

**Review by semicolin — CHANGES_REQUESTED** [review]
See questions below...

File: package.json (hunk 3 above)
Comment (nikhil-ht) [reply package.json]: @semicolin  that is correct, but i was face issue while publish lforms on npm  so i keep it form specified location.

File: app/views/partials/layout-grid.html (hunk 5 above)
Comment (nikhil-ht) [reply app/views/partials/layout-grid.html]: @semicolin  yes its creates additional column for each nested item . i haven't found such situation where each item has a different number of nested items  . and will remove that  misleading comment

File: package.json (hunk 3 above)
Comment (semicolin) [reply package.json]: It looks like the `build:mkdist` script creates a symlink for `dist/latest`. is that working correctly? I don't understand why it would work for the cleancss output file but not the input file.

File: package.json (hunk 4 above)
Comment (nikhil-ht) [reply package.json]: @semicolin I will check this and Update the PR accordingly, Will also update the description for same.

File: package.json (hunk 3 above)
Comment (nikhil-ht) [reply package.json]: Hi @semicolin Don't know how but the `$npm_package_name`  variable is referring to the correct package name, and that's why the issue occurs while creating directory to add css and other files. please check below ss.
It should create` test-0.0.5` dir and add all css the files in it as per the scripts. But strangely it is not working. Can you please check this at your side also ?
Screenshot from 2023-02-07 20-04-37

File: package.json (hunk 4 above)
Comment (hankwallace) [reply package.json]: @nikhil-ht Please don't let the etiquette here hold up this pr. You can research this and submit a follow-up pull request.

File: package.json (hunk 3 above)
Comment (semicolin) [reply package.json]: webpack.config.js line 55 has hardcoded package name. Change it to:
```js
let versionedDist = require('./package.json').name+'-'+require('./package.json').version;
```

Comment (nikhil-ht) [reply package.json]: okay @semicolin   I will update

File: package.json (hunk 4 above)
Comment (nikhil-ht) [reply package.json]: okay @hankwallace

**Review by semicolin — APPROVED** [review]
